### PR TITLE
Gamepad V1 is ready to go!

### PR DESCRIPTION
**Pull request by joncatanio**
This is the first version of the gamepad. It is pretty much solid, personally I love the movement but that might not reflect the average user. I did do some testing on two of my friends they really enjoyed the driving, they felt that it was unique and accurately represented good drifting mechanics. I should note that if you turn very very hard the user will go off screen. I haven't found an elegant solution for this yet but in the user testing I did nobody actually made that happen, and if they did they attributed it to "drifting" and didn't see it as a deterrence at all. **We need to fully draw the aiming arch length** this was the most asked feedback. The aiming with the right joystick is very smooth but needs to be made more obvious.

@nurgan I made some changes to the cookie action component and aim component, the joystick gives the ability to just use vectors straight up (you can see an example of this when I use the left joystick to drive). But when it comes to aiming I tried to use as much existing code as possible so it makes sense to you as to what I am doing. I think it is pretty straight forward and I didn't have to change many existing variables. It ended up working out pretty well but might not be as clean as possible (I think that is ok for this project).

### CODE REVIEW

File: src/AimInputComponent.cpp [hunk 1]
```diff
@@ -7,48 +7,101 @@
...
+}
+
+void AimInputComponent::pollGamepad() {
+   int count;
+   const float* axes = glfwGetJoystickAxes(GLFW_JOYSTICK_1, &count);
+   float xComponent = -axes[3], yComponent = -axes[2];
+   // A trigger at rest has a value of -1.0, fully engaged is 1.0.
+   float leftTrigger = axes[4], rightTrigger = axes[5];
+
+   if (-0.1f < xComponent && xComponent < 0.1f) {
```
Comment (rgarmsen2295) [inline src/AimInputComponent.cpp:34]: Guessing this is the "deadzone" check? Kinda cool to see it in code!

File: src/Camera.cpp [hunk 2]
```diff
@@ -74,6 +74,33 @@ void Camera::changeBeta(float deltaBeta) {
...
 
+/* Calculates the angle of the given direction and the original direction vector
+ * to determine the rotation angle to apply to the current direction vector.
+ * Note: The dot product doesn't specify sign therefore if the z component
+ * of the given vector is negative the angle will be negated.
+ * Note: If the player is in reverse the base direction for the angle
+ * calculation is reversed.
+ */
+void Camera::calcCamAndPlayerOrient(glm::vec3 dirOrigOrient, bool reverse) {
+   glm::vec3 origDirection, unitDir = glm::normalize(dirOrigOrient);
```
Comment (rgarmsen2295) [inline src/Camera.cpp]: Minor nit, but can you separate this into two lines? Tripped me up while reading it.

File: src/PlayerInputComponent.cpp [hunk 3]
```diff
@@ -5,15 +5,71 @@
...
+   bool modeChangeAvailable = !holder_->toggleMovement;
+
+   // If the joystick crosses the x-axis change driving mode.
+   if (xComponent == -1 || xComponent == 1) {
+      modeChangeAvailable = true;
+      inReverse = false;
+   }
+
+   // Clamp deadspace out so the character doesn't move around.
+   if (-0.1f < xComponent && xComponent < 0.1f &&
```
Comment (rgarmsen2295) [inline src/PlayerInputComponent.cpp:47]: More deadspace code : D

**Review by rgarmsen2295 — APPROVED** [review]
One super minor nitpick, otherwise looks great!  Just briefly glanced over the math since we will chat about tomorrow.

Also, make sure @nurgan reviews this before merge due to it interacting with his cookie code (want him to know about changes).

File: src/Camera.cpp (hunk 2 above)
Comment (joncatanio) [reply src/Camera.cpp]: Yep totally!

File: src/AimInputComponent.cpp (hunk 1 above)
Comment (joncatanio) [reply src/AimInputComponent.cpp:34]: Hahaha, Reed loving the dead space code XD. Yea I thought it was pretty cool myself haha, weird to see it!

File: src/AimInputComponent.cpp [hunk 4]
```diff
@@ -7,48 +7,101 @@
...
+// rotate direction with rotate if pressed
+void AimInputComponent::pollKeyboard() {
     if (glfwGetKey(window, GLFW_KEY_I) == GLFW_PRESS) {
         toggleXRotation = true;
-        rotationDirection = 1.0f;
+        rotationXDirection = 1.0f;
     } else if (glfwGetKey(window, GLFW_KEY_K) == GLFW_PRESS) {
         toggleXRotation = true;
-        rotationDirection = -1.0f;
+        rotationXDirection = -1.0f;
```
Comment (nurgan) [inline src/AimInputComponent.cpp:78]: yeah, that was more then neccesary to split in in x and y rotation

**Review by nurgan — APPROVED** [review]
I didn't look to much into the math, but the rest look very good

File: src/PlayerInputComponent.cpp [hunk 5]
```diff
@@ -5,15 +5,71 @@
...
+   GameManager& gameManager = GameManager::instance();
+   Camera& camera = gameManager.getCamera();
+   int count;
+   const float* axes = glfwGetJoystickAxes(GLFW_JOYSTICK_1, &count);
+
+   /* For PS4 and Xbox One controllers the axes count will be 6.
+    * Indices [0, 1] refer to the left joysticks x and y axes respectively.
+    * Indices [2, 3] refer to the right joysticks x and y axes respectively.
+    * Indices [4, 5] refer to the left and right triggers respectively.
+    * For our purposes x-joy = y-world & y-joy = z-world
```
Comment (joncatanio) [inline src/PlayerInputComponent.cpp]: x-joy goes to z-world && y-joy to x-world for driving

**Pull request merged**